### PR TITLE
Add sampler

### PR DESCRIPTION
**Pull request by cvetkovic**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,167 +1,21 @@
...
-
-For more options, please see the `Makefile`.
-
-## Using OpenWhisk
-
-For instructions on how to use the loader with OpenWhisk go to `openwhisk_setup/README.md`.
-
+# InVitro
+
+InVitro is a set of tools for conducting research in serverless computing. It consists of the sampler and the loader. The sampler is used for creating a representative workload summary from a real workload trace. The loader is a tool for generating the load based on a given trace, which is compatible with vHive and OpenWhisk.
```
Comment (ustiugov) [inline README.md:3]:
```suggestion
In-Vitro is a set of tools for analyzing the performance of serverless cluster deployments. In-Vitro consists of two tools, namely sampler and loader. Sampler creates representative workload summaries (i.e., samples of functions) based on production traces. Loader reconstructs the invocation traffic based on a given trace and steers this load to the functions deployed in the studied serverless cluster. Currently, In-Vitro supports [vHive](LINK) and [OpenWhisk](LINK).
```

provide links to vHive and OpenWhisk

File: README.md [hunk 2]
```diff
@@ -1,167 +1,21 @@
...
+
```
Comment (ustiugov) [inline README.md:4]: this doc needs to have the following subsections:
- license, etc (see vHive)
- Pointers to the quickstart and other first-order docs

File: docs/loader.md [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+The experiments require a server-grade node running Linux (tested on Ubuntu 20, Intel Xeon). On CloudLab, one
+can choose the APT cluster `d430` node.
+
+## Create a cluster
+
+First, configure `script/setup/setup.cfg`. You can specify there which vHive branch to use, loader branch, operation mode
+(sandbox type), maximum number of pods per node, and the Github token. All these configurations are mandatory.
+We currently support the following modes: containerd (`container`), Firecracker (`firecracker`), and Firecracker with
+snapshots (`firecracker_snapshots`).
+The token needs `repo` and `admin:public_key` permissions and will be used for adding SSH key to the user's account for
```
Comment (ustiugov) [inline docs/loader.md:16]: now all repos are public but you may keep this note in the bottom of the doc

File: .github/workflows/preprocessing_tests.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      - main
+    paths:
+      - 'sampler/preprocess.py'
+      - 'sampler/tests/preprocess_test.py'
+      - 'sampler/__main__.py'
+  workflow_dispatch:
+
+jobs:
+  preprocessing:
+    name: Preprocessing tests
```
Comment (ustiugov) [inline .github/workflows/preprocessing_tests.yaml:22]:
```suggestion
    name: Sampler preprocessing tests
```

**Comment by ustiugov** [comment]
Please give this one a priority. We need to start advertising the work ASAP

File: docs/loader.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+## Tune the timing for the benchmark function
+
+Before start any experiments, the timing of the benchmark function should be tuned so that it consumes the required
+service time more precisely.
+
+First, run the following command to deploy the timing benchmark that yields the number of execution units* the function
+needs to run given a required service time.
+
+* The execution unit is approximately 100 `SQRTSD` x86 instructions.
```
Comment (ustiugov) [inline docs/loader.md:94]: can we get this replaced by a simple albeit cross-platform counter increment? please raise an issue

File: README.md [hunk 6]
```diff
@@ -1,184 +1,43 @@
...
 
-For instructions on how to use the loader with OpenWhisk go to `openwhisk_setup/README.md`.
+## License and copyright
 
-## Running on Cloud Using Serverless Framework
+InVitro is free. We publish the code under the terms of the MIT License that allows distribution, modification, and commercial use.
+This software, however, comes without any warranty or liability.
 
-Currently supported vendors: AWS
+The software is maintained at the [EASL lab](https://systems.ethz.ch/research/easl.html) at ETH Zürich.
```
Comment (ustiugov) [inline README.md:38]:
```suggestion
The software is maintained by the [EASL lab](https://systems.ethz.ch/research/easl.html) at ETH Zürich.
```

**Review by ustiugov — CHANGES_REQUESTED** [review]
a couple of comments, otherwise lgtm

**Comment by ustiugov** [comment]
@leokondrashov please provide your review as well. particularly, check the READMEs and how clear they are

File: README.md [hunk 7]
```diff
@@ -1,184 +1,42 @@
...
 
-## Using OpenWhisk
+We would be happy to answer any questions in GitHub Issues and encourage the open-source community
+to submit new Issues, assist in addressing existing issues and limitations, and contribute their code with Pull Requests.
+Please check our guide on [Contributing to vHive](https://github.com/vhive-serverless/vHive/docs/contributing_to_vhive.md) if you would like contribute.
+You can also talk to us in our [Slack space](https://join.slack.com/t/vhivetutorials/shared_invite/zt-1fk4v71gn-nV5oev5sc9F4fePg3_OZMQ).
 
-For instructions on how to use the loader with OpenWhisk go to `openwhisk_setup/README.md`.
 
-## Running on Cloud Using Serverless Framework
```
Comment (leokondrashov) [inline README.md:168]: This part got lost in movement to new docs.

File: docs/sampler.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+invocation duration average is 1000ms, invocation count average is <10/min, and memory average is 200MB.
+
+First, the sampler derives the largest sample, for which it makes several attempts (trials) when choosing the sample
+with the smallest aggregate WD distance (mean of the invocation and resource WDs). Then, each smaller sample is derived
+from the previous, larger sample; hence guaranteeing that large samples always include smaller samples that guarantees
+monotonic load increase (in terms of resource usage) when sweeping the sample size.
+
+## Getting the traces
+
+The reference traces are stored in the `reference_traces` branch of this repo in the root folder, as a `samples.tar.gz`
```
Comment (leokondrashov) [inline docs/sampler.md:86]: This refers to the sampler branch. Will this be added as well or should we just add it to this repo's LFS?

File: docs/sampler.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                        Step (#functions) in sample size during sampling.
+  -max integer, --max-size integer
+                        Maximum sample size (#functions).
+  -tr integer, --trial-num integer
+                        Number of sampling trials for each sample size.
+```
+
+### Plotting
+
+**Note:** Currently plotting is broken and has been disabled. (Issue 77)
```
Comment (leokondrashov) [inline docs/sampler.md:115]: As well, referencing the Issue from the old sampler repo. Should we transfer issues?

**Review by leokondrashov — CHANGES_REQUESTED** [review]
Looked over READMEs, left couple comments.

**Pull request merged**